command/agent/auth/auth.go

```diff
@@ -212,6 +216,20 @@ func (ah *AuthHandler) Run(ctx context.Context, am AuthMethod) error {
 			}
 		}
 
+		clientAddr = clientToUse.Address()
+		tlsServerName = clientToUse.CloneConfig().HttpClient.Transport.(*http.Transport).TLSClientConfig.ServerName
+		if tlsServerName != "" && tlsServerName != clientAddr {
```

Comment:
There's no requirement that tls server name matches the address we use, which e.g. could be an IP address.